From 5ae45eaa7e8de79e1e6c542b1accb42d47ce9ce7 Mon Sep 17 00:00:00 2001
From: Won-Kyu Park <wkpark@gmail.com>
Date: Tue, 24 Jan 2012 22:35:33 +0900
Subject: [PATCH] Support screenshot / renderScreenToTexture under OpenGL ES
 1.0

current implemented screenshot / renderScreenToTexture() of
SurfaceFlinger use FrameBuffer Object(FBO) in the OpenGL ES 1.1/2.0
extention

directly use framebuffer device /dev/graphics/fb0 to emulate FBO.

Tested under the old adreno130/msm7201 (unofficial CM9)

Please see also http://review.cyanogenmod.com/11022

Change-Id: I0eada5bfe9b389f491691b770f5662935d0fe8e4
---
 services/surfaceflinger/SurfaceFlinger.cpp |  266 +++++++++++++++++++++++++++-
 services/surfaceflinger/SurfaceFlinger.h   |    9 +
 2 files changed, 270 insertions(+), 5 deletions(-)

diff --git a/services/surfaceflinger/SurfaceFlinger.cpp b/services/surfaceflinger/SurfaceFlinger.cpp
index e3a009e..ffd09a7 100644
--- a/services/surfaceflinger/SurfaceFlinger.cpp
+++ b/services/surfaceflinger/SurfaceFlinger.cpp
@@ -26,6 +26,7 @@
 #include <limits.h>
 #include <sys/types.h>
 #include <sys/stat.h>
+#include <linux/fb.h>
 #include <sys/ioctl.h>
 
 #include <cutils/log.h>
@@ -1880,7 +1881,7 @@ status_t SurfaceFlinger::renderScreenToTextureLocked(DisplayID dpy,
     ATRACE_CALL();
 
     if (!GLExtensions::getInstance().haveFramebufferObject())
-        return INVALID_OPERATION;
+        return directRenderScreenToTextureLocked(dpy, textureName, uOut, vOut);
 
     // get screen geometry
     const DisplayHardware& hw(graphicPlane(dpy).displayHardware());
@@ -1940,6 +1941,144 @@ status_t SurfaceFlinger::renderScreenToTextureLocked(DisplayID dpy,
     return NO_ERROR;
 }
 
+
+status_t SurfaceFlinger::directRenderScreenToTextureLocked(DisplayID dpy,
+        GLuint* textureName, GLfloat* uOut, GLfloat* vOut)
+{
+    status_t result;
+    const DisplayHardware& hw(graphicPlane(dpy).displayHardware());
+
+    // use device framebuffer in /dev/graphics/fb0
+    size_t offset;
+    uint32_t bytespp, format, gl_format, gl_type;
+    size_t size = 0;
+    struct fb_var_screeninfo vinfo;
+    const char* fbpath = "/dev/graphics/fb0";
+    int fb = open(fbpath, O_RDONLY);
+    void const* mapbase = MAP_FAILED;
+    ssize_t mapsize = -1;
+
+    if (fb < 0) {
+        LOGE("Failed to open framebuffer");
+        return INVALID_OPERATION;
+    }
+
+    if (ioctl(fb, FBIOGET_VSCREENINFO, &vinfo) < 0) {
+        LOGE("Failed to get framebuffer info");
+        close(fb);
+        return INVALID_OPERATION;
+    }
+
+    bytespp = vinfo.bits_per_pixel / 8;
+    const uint32_t hw_w = vinfo.xres;
+    const uint32_t hw_h = vinfo.yres;
+
+    switch (bytespp) {
+    case 2:
+        format = PIXEL_FORMAT_RGB_565;
+        gl_format = GL_RGB;
+        gl_type = GL_UNSIGNED_SHORT_5_6_5;
+        break;
+    case 4:
+        format = PIXEL_FORMAT_RGBX_8888;
+        gl_format = GL_RGBA;
+        gl_type = GL_UNSIGNED_BYTE;
+        break;
+    default:
+        close(fb);
+        LOGE("Failed to detect framebuffer bytespp");
+        return INVALID_OPERATION;
+        break;
+    }
+
+    offset = (vinfo.xoffset + vinfo.yoffset * vinfo.xres) * bytespp;
+    size = vinfo.xres * vinfo.yres * bytespp;
+
+    mapsize = offset + size;
+    mapbase = mmap(0, mapsize, PROT_READ, MAP_PRIVATE, fb, 0);
+    close(fb);
+    if (mapbase == MAP_FAILED) {
+        return INVALID_OPERATION;
+    }
+
+    void const* fbbase = (void *)((char const *)mapbase + offset);
+    GLfloat u = 1;
+    GLfloat v = 1;
+
+    // build texture
+    GLuint tname;
+    glGenTextures(1, &tname);
+    glBindTexture(GL_TEXTURE_2D, tname);
+    glTexImage2D(GL_TEXTURE_2D, 0, gl_format,
+            hw_w, hw_h, 0, gl_format, GL_UNSIGNED_BYTE, 0);
+    if (glGetError() != GL_NO_ERROR) {
+        while ( glGetError() != GL_NO_ERROR ) ;
+        GLint tw = (2 << (31 - clz(hw_w)));
+        GLint th = (2 << (31 - clz(hw_h)));
+        glTexImage2D(GL_TEXTURE_2D, 0, gl_format,
+                tw, th, 0, gl_format, GL_UNSIGNED_BYTE, 0);
+        u = GLfloat(hw_w) / tw;
+        v = GLfloat(hw_h) / th;
+    }
+
+    // write fb data to image buffer texture (reverse order)
+    GLubyte* imageData = (GLubyte*)malloc(size);
+    if (imageData) {
+        void *ptr = imageData;
+        uint32_t rowlen = hw_w * bytespp;
+        offset = size;
+        for (uint32_t j = hw_h; j > 0; j--) {
+            offset -= rowlen;
+            memcpy(ptr, fbbase + offset, rowlen);
+            ptr += rowlen;
+        }
+
+        // write image buffer to the texture
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+
+        // copy imageData to the texture
+        glTexImage2D(GL_TEXTURE_2D, 0, gl_format, hw_w, hw_h, 0,
+                gl_format, gl_type, imageData);
+
+        LOGI("direct Framebuffer texture for gl_format=%d gl_type=%d", gl_format, gl_type);
+        result = NO_ERROR;
+    } else {
+        result = NO_MEMORY;
+    }
+
+    // redraw the screen entirely...
+    glDisable(GL_TEXTURE_2D);
+    glDisable(GL_SCISSOR_TEST);
+    glClearColor(0,0,0,1);
+    glClear(GL_COLOR_BUFFER_BIT);
+    glEnable(GL_SCISSOR_TEST);
+    glMatrixMode(GL_MODELVIEW);
+    glLoadIdentity();
+    const Vector< sp<LayerBase> >& layers(mVisibleLayersSortedByZ);
+    const size_t count = layers.size();
+    for (size_t i=0 ; i<count ; ++i) {
+        const sp<LayerBase>& layer(layers[i]);
+        layer->drawForSreenShot();
+    }
+
+    hw.compositionComplete();
+
+    // done
+    munmap((void *)mapbase, mapsize);
+
+    *textureName = tname;
+    *uOut = u;
+    *vOut = v;
+
+    // free buffer memory
+    if (imageData) {
+        free(imageData);
+    }
+
+    return result;
+}
+
 // ---------------------------------------------------------------------------
 
 class VSyncWaiter {
@@ -2407,6 +2546,125 @@ status_t SurfaceFlinger::turnElectronBeamOn(int32_t mode)
 
 // ---------------------------------------------------------------------------
 
+status_t SurfaceFlinger::directCaptureScreenImplLocked(DisplayID dpy,
+        sp<IMemoryHeap>* heap,
+        uint32_t* w, uint32_t* h, PixelFormat* f,
+        uint32_t sw, uint32_t sh,
+        uint32_t minLayerZ, uint32_t maxLayerZ)
+{
+    status_t result = PERMISSION_DENIED;
+
+    uint32_t width, height, format;
+    uint32_t bytespp;
+    void const* mapbase = MAP_FAILED;
+    ssize_t mapsize = -1;
+
+    struct fb_var_screeninfo vinfo;
+    const char* fbpath = "/dev/graphics/fb0";
+
+    // only one display supported for now
+    if (UNLIKELY(uint32_t(dpy) >= DISPLAY_COUNT))
+        return BAD_VALUE;
+
+    // get screen geometry
+    const DisplayHardware& hw(graphicPlane(dpy).displayHardware());
+    const uint32_t hw_w = hw.getWidth();
+    const uint32_t hw_h = hw.getHeight();
+
+    if ((sw > hw_w) || (sh > hw_h))
+        return BAD_VALUE;
+
+    sw = (!sw) ? hw_w : sw;
+    sh = (!sh) ? hw_h : sh;
+
+    int fb = open(fbpath, O_RDONLY);
+    if (fb < 0) {
+        LOGE("Failed to open framebuffer");
+        return INVALID_OPERATION;
+    }
+
+    if (ioctl(fb, FBIOGET_VSCREENINFO, &vinfo) < 0) {
+        LOGE("Failed to get framebuffer info");
+        close(fb);
+        return INVALID_OPERATION;
+    }
+
+    bytespp = vinfo.bits_per_pixel / 8;
+    switch (bytespp) {
+    case 2:
+        format = PIXEL_FORMAT_RGB_565;
+        break;
+    case 4:
+        format = PIXEL_FORMAT_RGBX_8888;
+        break;
+    default:
+        close(fb);
+        LOGE("Failed to detect framebuffer bytespp");
+        return INVALID_OPERATION;
+        break;
+    }
+
+    size_t offset = (vinfo.xoffset + vinfo.yoffset * vinfo.xres) * bytespp;
+    size_t size = vinfo.xres * vinfo.yres * bytespp;
+    mapsize = offset + size;
+    mapbase = mmap(0, mapsize, PROT_READ, MAP_PRIVATE, fb, 0);
+    close(fb);
+    if (mapbase == MAP_FAILED) {
+        return INVALID_OPERATION;
+    }
+
+    width = vinfo.xres;
+    height = vinfo.yres;
+    void const* fbbase = (void *)((char const *)mapbase + offset);
+
+    const LayerVector& layers(mDrawingState.layersSortedByZ);
+    const size_t count = layers.size();
+    for (size_t i=0 ; i<count ; ++i) {
+        const sp<LayerBase>& layer(layers[i]);
+        const uint32_t flags = layer->drawingState().flags;
+        if (!(flags & ISurfaceComposer::eLayerHidden)) {
+            const uint32_t z = layer->drawingState().z;
+            if (z >= minLayerZ && z <= maxLayerZ) {
+                layer->drawForSreenShot();
+            }
+        }
+    }
+
+    // allocate shared memory large enough to hold the
+    // screen capture
+    size = sw * sh * bytespp;
+    sp<MemoryHeapBase> heapBase(
+            new MemoryHeapBase(size, 0, "screen-capture") );
+    void* ptr = heapBase->getBase();
+
+    if (ptr) {
+        if ((sw == hw_w) && (sh == hw_h)) {
+            memcpy(ptr, fbbase, size);
+        } else {
+            uint32_t rowlen = hw_w * bytespp;
+            uint32_t collen = sw * bytespp;
+            size_t offset = 0;
+            for (uint32_t j = 0; j < sh; j++) {
+                memcpy(ptr, fbbase + offset, collen);
+                ptr += collen;
+                offset += rowlen;
+            }
+        }
+        *heap = heapBase;
+        *w = sw;
+        *h = sh;
+        *f = format;
+        result = NO_ERROR;
+    } else {
+        result = NO_MEMORY;
+    }
+    munmap((void *)mapbase, mapsize);
+
+    hw.compositionComplete();
+
+    return result;
+}
+
 status_t SurfaceFlinger::captureScreenImplLocked(DisplayID dpy,
         sp<IMemoryHeap>* heap,
         uint32_t* w, uint32_t* h, PixelFormat* f,
@@ -2422,7 +2680,8 @@ status_t SurfaceFlinger::captureScreenImplLocked(DisplayID dpy,
         return BAD_VALUE;
 
     if (!GLExtensions::getInstance().haveFramebufferObject())
-        return INVALID_OPERATION;
+        return directCaptureScreenImplLocked(dpy,
+                heap, w, h, f, sw, sh, minLayerZ, maxLayerZ);
 
     // get screen geometry
     const DisplayHardware& hw(graphicPlane(dpy).displayHardware());
@@ -2538,9 +2797,6 @@ status_t SurfaceFlinger::captureScreen(DisplayID dpy,
     if (CC_UNLIKELY(uint32_t(dpy) >= DISPLAY_COUNT))
         return BAD_VALUE;
 
-    if (!GLExtensions::getInstance().haveFramebufferObject())
-        return INVALID_OPERATION;
-
     class MessageCaptureScreen : public MessageBase {
         SurfaceFlinger* flinger;
         DisplayID dpy;
diff --git a/services/surfaceflinger/SurfaceFlinger.h b/services/surfaceflinger/SurfaceFlinger.h
index 0492aa3..0fe0c66 100644
--- a/services/surfaceflinger/SurfaceFlinger.h
+++ b/services/surfaceflinger/SurfaceFlinger.h
@@ -355,6 +355,15 @@ private:
                     uint32_t reqWidth, uint32_t reqHeight,
                     uint32_t minLayerZ, uint32_t maxLayerZ);
 
+            status_t directCaptureScreenImplLocked(DisplayID dpy,
+                    sp<IMemoryHeap>* heap,
+                    uint32_t* width, uint32_t* height, PixelFormat* format,
+                    uint32_t reqWidth, uint32_t reqHeight,
+                    uint32_t minLayerZ, uint32_t maxLayerZ);
+
+            status_t directRenderScreenToTextureLocked(DisplayID dpy,
+                    GLuint* textureName, GLfloat* uOut, GLfloat* vOut);
+
             status_t turnElectronBeamOffImplLocked(int32_t mode);
             status_t turnElectronBeamOnImplLocked(int32_t mode);
             status_t electronBeamOffAnimationImplLocked();
-- 
1.7.9.5

